src/core/ext/transport/chaotic_good/client_transport.h

```diff
@@ -129,30 +149,44 @@ class ClientTransport {
                   // Save incomming frame results to call_args.
                   [server_initial_metadata, server_to_client_messages](
                       absl::optional<ServerFrame> server_frame) mutable {
-                    GPR_ASSERT(server_frame.has_value());
-                    auto frame = std::move(
-                        absl::get<ServerFragmentFrame>(*server_frame));
+                    bool transport_closed = false;
+                    ServerFragmentFrame frame;
+                    if (!server_frame.has_value()) {
+                      // Incoming server frame pipe is closed, which only
+                      // happens when transport is aborted.
+                      transport_closed = true;
+                    } else {
+                      frame = std::move(
+                          absl::get<ServerFragmentFrame>(*server_frame));
+                    };
                     return TrySeq(
-                        If((frame.headers != nullptr),
+                        If((!transport_closed) && (frame.headers != nullptr),
                            [server_initial_metadata,
                             headers = std::move(frame.headers)]() mutable {
                              return server_initial_metadata->Push(
                                  std::move(headers));
                            },
                            [] { return false; }),
-                        If((frame.message != nullptr),
+                        If((!transport_closed) && (frame.message != nullptr),
                            [server_to_client_messages,
                             message = std::move(frame.message)]() mutable {
                              return server_to_client_messages->Push(
                                  std::move(message));
                            },
                            [] { return false; }),
-                        If((frame.trailers != nullptr),
+                        If((!transport_closed) && (frame.trailers != nullptr),
                            [trailers = std::move(frame.trailers)]() mutable
                            -> LoopCtl<ServerMetadataHandle> {
                              return std::move(trailers);
                            },
-                           []() -> LoopCtl<ServerMetadataHandle> {
+                           [transport_closed]()
+                               -> LoopCtl<ServerMetadataHandle> {
+                             if (transport_closed) {
+                               return ServerMetadataFromStatus(
+                                   absl::UnavailableError(
+                                       "Transport closed due to endpoint "
```

Comment:
We should figure out a way to get the actual error message on why the read/write failed. Pls add a TODO

Reply:
Done, thanks for the suggestion.